File: C/0117-填充每个节点的下一个右侧节点指针 II.c
// Definition for a Node.
#include <cstddef>
static struct Node {
	int val;
	struct Node* left;
	struct Node* right;
	struct Node* next;
};

static struct Node* connect(struct Node* root) {
	if (!root) {
		return NULL;
	}
	struct Node* q[10001];
	int left = 0, right = 0;
	q[right++] = root;
	while (left < right) {
		int n = right - left;
		struct Node* last = NULL;
		for (int i = 1; i <= n; ++i) {
			struct Node* f = q[left++];
			if (f->left) {
				q[right++] = f->left;
			}
			if (f->right) {
				q[right++] = f->right;
			}
			if (i != 1) {
				last->next = f;
			}
			last = f;
		}
	}
	return root;
}

File: JavaScript/0117-填充每个节点的下一个右侧节点指针 II.js
// Definition for a Node.
function Node(val, left, right, next) {
	this.val = val === undefined ? null : val;
	this.left = left === undefined ? null : left;
	this.right = right === undefined ? null : right;
	this.next = next === undefined ? null : next;
}

// 和116没有任何差别

/**
 * @param  {Node} root
 * @return {Node}
 */
function connect(root) {
	if (root === null) {
		return root;
	}
	const queue = [root];
	while (queue.length !== 0) {
		const length = queue.length;
		for (let i = 0; i < length; i++) {
			const node = queue.shift();
			if (i < length - 1) {
				node.next = queue[0];
			}
			node.left && queue.push(node.left);
			node.right && queue.push(node.right);
		}
	}
	return root;
}
